### PR TITLE
Validate opening and closing brackets on Anatomy keys.

### PR DESCRIPTION
**Pull request by robin-ynput**
## Changelog Description

Additional fix to address: https://github.com/ynput/ayon-nuke/issues/9 root cause
Currently anatomy keys with unmached `[` and `]` will be hard failing at formatting attempt. Redirect to `AnatomyTemplateUnsolved` instead. 

## Additional info

@iLLiCiTiT this one is mostly to trigger a discussion. See how you like it, I might not have fixed the way it should (hence keeping it as draft). Please feel free to recommend another implementation.

## Testing notes:

Introduce a `variant]` key in the publish/test/directory anatomy setting.

```python
from ayon_core.pipeline import Anatomy, get_current_project_name

prj = get_current_project_name()
ana = Anatomy(prj)
ttt = ana.format(data)
ttt["publish"]["test"]["directory"]
```

before fix:
``` text
  File "C:\Users\robert\AppData\Local\Ynput\AYON\addons\core_0.4.2\ayon_core\pipeline\anatomy\templates.py", line 348, in format
    value = value.format(data, strict)
  File "C:\Users\robert\AppData\Local\Ynput\AYON\addons\core_0.4.2\ayon_core\pipeline\anatomy\templates.py", line 297, in format
    value = value.format(data)
  File "C:\Users\robert\AppData\Local\Ynput\AYON\addons\core_0.4.2\ayon_core\pipeline\anatomy\templates.py", line 96, in format
    result = StringTemplate.format(self, data)
  File "C:\Users\robert\AppData\Local\Ynput\AYON\addons\core_0.4.2\ayon_core\lib\path_templates.py", line 117, in format
    part.format(data, result)
  File "C:\Users\robert\AppData\Local\Ynput\AYON\addons\core_0.4.2\ayon_core\lib\path_templates.py", line 532, in format
    formatted_value = self.template.format(**fill_data)
KeyError: 'variant]'
```

after fix:
```text
raise AnatomyTemplateUnsolved(ayon_core.pipeline.anatomy.exceptions.AnatomyTemplateUnsolved: Anatomy template "{root[work]}/{project[name]}/{hierarchy}/{folder[name]}/source/{variant]}/v{version:0>3}" is unsolved. Missing keys: "variant]". Keys with invalid data type: `"version" <class 'NoneType'>`.
```

### CODE REVIEW

File: client/ayon_core/lib/path_templates.py [hunk 1]
```diff
@@ -460,6 +460,30 @@ def validate_value_type(value):
                 return True
         return False
 
+    @staticmethod
+    def validate_key_is_matched(key):
```
Comment (iLLiCiTiT) [inline client/ayon_core/lib/path_templates.py:464]: I'm afraid we don't support inner keys in template (e.g. `{{{some[key]}}}`), so we don't need to validate it, validating that is out of scope of this PR. We also don't support `()`. So only remaining check is `[]`.

But because I like it, we might just mention it in docstring.

File: client/ayon_core/lib/path_templates.py [hunk 2]
```diff
@@ -460,6 +460,30 @@ def validate_value_type(value):
...
+        Finds out how balanced an expression is.
+        With a string containing only brackets.
+
+        >>> is_matched('[]()()(((([])))')
+        False
+        >>> is_matched('[](){{{[]}}}')
+        True
+        """
+        opening = tuple('({[')
+        closing = tuple(')}]')
```
Comment (iLLiCiTiT) [inline client/ayon_core/lib/path_templates.py:475]: Double quotes are preferred.
```suggestion
        opening = tuple("({[")
        closing = tuple(")}]")
```

Comment (robin-ynput) [reply client/ayon_core/lib/path_templates.py:478]: Noted !

File: client/ayon_core/lib/path_templates.py (hunk 1 above)
Comment (robin-ynput) [reply client/ayon_core/lib/path_templates.py:464]: Adjusted, thanks for the feedback !

File: client/ayon_core/lib/path_templates.py [hunk 3]
```diff
@@ -460,6 +460,34 @@ def validate_value_type(value):
...
+            >>> is_matched("[](){{{[]}}}")
+            True
+
+        Returns:
+            bool: Openings and closing are valid.
+
+        """
+        opening = tuple("({[")
+        closing = tuple(")}]")
+        mapping = dict(zip(opening, closing))
```
[the remark this answers is not in the document]
Comment (robin-ynput) [reply client/ayon_core/lib/path_templates.py:480]: be picky please !

**Comment by BigRoy** [comment]
Out of curiosity - after the fix this reports:

> `AnatomyTemplateUnsolved(ayon_core.pipeline.anatomy.exceptions.AnatomyTemplateUnsolved: Anatomy template "{root[work]}/{project[name]}/{hierarchy}/{folder[name]}/source/{variant]}/v{version:0>3}" is unsolved. Missing keys: "variant]". Keys with invalid data type: `"version" <class 'NoneType'>``

Where does that invalid data type of version come from in that report? Seems... unrelated or oddly reported together with the missing key?

**Review by iLLiCiTiT — APPROVED** [review]

**Comment by iLLiCiTiT** [comment]
> Where does that invalid data type of version come from in that report? Seems... unrelated or oddly reported together with the missing key?

What was the template?

**Comment by BigRoy** [comment]
> What was the template? Or where did you came up with data you've used?

If you're asking me. I took this line from @robin-ynput PR description :)

**Comment by robin-ynput** [comment]
`{root[work]}/{project[name]}/{hierarchy}/{folder[name]}/source/{variant]}/{@version}` in `publish/test/directory`

**Comment by BigRoy** [comment]
> @BigRoy @iLLiCiTiT `{root[work]}/{project[name]}/{hierarchy}/{folder[name]}/source/{variant]}/{@version}` in `publish/test/directory`

I see now. It's just a report of all the invalid things after each other, one one line. It's a bit confusing to read without context but I guess in a way it makes sense? Fine to ignore for me unless we (or just me?) get confused later again.

**Review by BigRoy — APPROVED** [review]
